File: credit.c
//Check credit card number against Luhn's algorithm
#include <cs50.h>
#include <stdio.h>

int main(void)
{
    //Ask for card number make sure number is positive
    long long cnum;
    do
    {
        cnum = get_long_long("Number: ");
    }
    while (cnum < 0);


    //Declare and define card type variables
    string ctype = "INVALID";
    char part_check = 'n';

    //Declare and define Luhn's algorithm variables
    int digit;
    int digit_length = 0;
    int evn_sum = 0;
    int odd_sum = 0;

    //Iterate through each digit from last to first, break for too many digits
    bool too_long = false;
    while (cnum > 0 && too_long == false)
    {
        digit = cnum % 10;
        cnum -= digit;
        cnum /= 10;
        digit_length++;

        //Check final digits for card type, or break if too many digits
        if (digit_length > 12)
        {
            if (digit_length == 13 && digit == 4)
            {
                ctype = "VISA";
            }
            else if (digit_length == 14)
            {
                if (digit == 4 || digit == 7)
                {
                    part_check = 'a';
                }
                else
                {
                    ctype = "INVALID";
                }
            }
            else if (digit_length == 15)
            {
                if (digit == 3 && part_check == 'a')
                {
                    ctype = "AMEX";
                }
                else if (digit >= 1 && digit <= 5)
                {
                    part_check = 'm';
                }
            }
            else if (digit_length == 16)
            {
                if (digit == 5 && part_check == 'm')
                {
                    ctype = "MASTERCARD";
                }
                else if (digit == 4)
                {
                    ctype = "VISA";
                }
                else
                {
                    ctype = "INVALID";
                    too_long = true;
                }
            }
        }

        //Apply Luhn's Algorithm
        if (digit_length % 2 == 0)
        {
            //Even digits: double and sum digits
            digit *= 2;
            evn_sum += digit % 10;
            digit -= digit % 10;
            digit /= 10;
            evn_sum += digit % 10;
        }
        else
        {
            //Odd digits: sum
            odd_sum += digit;
        }

    }

    if ((evn_sum + odd_sum) % 10 != 0)
    {
        ctype = "INVALID";
    }
    printf("%s\n", ctype);

}


//Define Functions

File: credit.py
"""Credit Homework Problem"""
from cs50 import get_int


def main():
    """Determine card type and validty of card using Luhn's algorithm."""

    # Obtain users card number
    cNum = get_int("Number: ")

    # Ensure users input is valid card number
    while cNum < 0:
        cNum = get_int("Number: ")

    # Initiate varaible to hold sum and convert card number to a string
    total = 0
    cNum = str(cNum)

    # Iterate over every other digit starting from the penultimate
    for dLoc in range(len(cNum) - 1, 0, -2):

        # Retrive digit from card number
        cDigit = int(cNum[dLoc - 1])

        # Multiply card's digit by 2 and convert to string
        product = str(cDigit * 2)

        # Iterate over each digit of product
        for pDigit in product:

            # Add product's digit to total
            total += int(pDigit)

    # Iterate over every other digit starting from the last
    for dLoc in range(len(cNum), 0, -2):

        # Retrive digit from card number
        digit = int(cNum[dLoc - 1])

        # Add digit total
        total += digit

    # Initive varaible to store validity of card
    valid = False

    # Check card number is valid
    if (total % 10) == 0:

        # Initiate dictionaries to hold valid card types
        cNumLens = {13: "V", 15: "A", 16: "M/V"}
        cTypes = {'A': ["34", "37"],
                  'M': ["51", "52", "53", "54", "55"],
                  'V': ["4"]}

        # Determine card type from card number's length
        cType = cNumLens[len(cNum)]

        # Check if possible to be American Express
        if 'A' in cType:

            # Check if valid American Express
            if cNum[:2] in cTypes['A']:

                # Set validty to true and print card type
                valid = True
                print("AMEX")

        # Check if possible to be Mastercard
        if 'M' in cType:

            # Check if valid Mastercard
            if cNum[:2] in cTypes['M']:

                # Set validty to true and print card type
                valid = True
                print("MASTERCARD")

        # Check if possible to be Visa
        if 'V' in cType:

            # Check if valid Visa
            if cNum[:1] in cTypes['V']:

                # Set validty to true and print card type
                valid = True
                print("VISA")

    # Check if card is not valid
    if not valid:

        # Print card is invalid
        print("INVALID")

    # Return 0 if successful
    return 0


if __name__ == "__main__":
    # If program is run call main
    main()